tiledb/array.py

```diff
@@ -243,3 +251,761 @@ def index_domain_subarray(array, dom, idx: tuple):
+        if key:
+            config = tiledb.Config()
+            config["sm.encryption_key"] = key
+            config["sm.encryption_type"] = "AES_256_GCM"
+            ctx = tiledb.Ctx(config=config)
+        else:
+            ctx = default_ctx()
+
+    _mode_to_query_type = {
+        "r": lt.QueryType.READ,
+        "w": lt.QueryType.WRITE,
+        "m": lt.QueryType.MODIFY_EXCLUSIVE,
+        "d": lt.QueryType.DELETE,
+    }
+    if mode not in _mode_to_query_type:
+        raise ValueError("TileDB array mode must be 'r', 'w', 'm', or 'd'")
+    query_type = _mode_to_query_type[mode]
+
+    ts_start = None
+    ts_end = None
+    if timestamp is not None:
+        if isinstance(timestamp, tuple):
+            if len(timestamp) != 2:
+                raise ValueError(
+                    "'timestamp' argument expects either int or tuple(start: int, end: int)"
+                )
+            if timestamp[0] is not None:
+                ts_start = timestamp[0]
+            if timestamp[1] is not None:
+                ts_end = timestamp[1]
+        elif isinstance(timestamp, int):
+            # handle the existing behavior for unary timestamp
+            # which is equivalent to endpoint of the range
+            ts_end = timestamp
+        else:
+            raise TypeError("Unexpected argument type for 'timestamp' keyword argument")
+
+    return lt.Array(ctx, uri, query_type, (ts_start, ts_end))
+
+
+class Array:
+    """Base class for TileDB array objects.
+
+    Defines common properties/functionality for the different array types. When
+    an Array instance is initialized, the array is opened with the specified mode.
+
+    :param str uri: URI of array to open
+    :param str mode: (default 'r') Open the array object in read 'r', write 'w', or delete 'd' mode
+    :param str key: (default None) If not None, encryption key to decrypt the array
+    :param tuple timestamp: (default None) If int, open the array at a given TileDB
+        timestamp. If tuple, open at the given start and end TileDB timestamps.
+    :param str attr: (default None) open one attribute of the array; indexing a
+        dense array will return a Numpy ndarray directly rather than a dictionary.
+    :param Ctx ctx: TileDB context
+    """
+
+    def __init__(
+        self, uri, mode="r", key=None, timestamp=None, attr=None, ctx=None, **kwargs
+    ):
+        if not ctx:
+            ctx = default_ctx()
+
+        if "preloaded_array" not in kwargs:
+            self.array = preload_array(uri, mode, key, timestamp, ctx)
+        else:
+            self.array = kwargs["preloaded_array"]
+
+        # view on a single attribute
+        schema = self.array._schema()
+        if attr and not any(
+            attr == schema._attr(i)._name for i in range(schema._nattr)
+        ):
+            self.array._close()
+            raise KeyError("No attribute matching '{}'".format(attr))
+        else:
+            self.view_attr = attr if attr is not None else None
```

Comment:
`None` check is superflorous.

```suggestion
            self.view_attr = attr
```